### ISSUE TITLE
Engine Discovery 

### ISSUE THREAD

**Issue by nathanhammond** [dgeb/ember-engines#99]
A feature that numerous people are discussing is to have a host application be able to "discover" available engines to lazy load. This is a multi-step process:

1. We must define an engine bundle format.
2. We must support lazily-loaded engines that are fully-defined at build time with the client already aware of the engines.
3. We must define a way to advertise available engines.
4. We must create a way to inject the advertised engines into the host application.

Known issues that must be solved:
* Deep linking (currently solved at build time by propagating all routes to the top).
* How should this play with the `{{mount}}` keyword?
* How should this handle nested engines?
* ???

Moving toward a solution to this problem should focus initially on getting us to lazy-loading, but we should keep this in the back of our minds so as not to paint ourselves into a corner.

/cc @mike183 @ebryn @miguelmadero @pestoney @marcemira @trentmwillis

**Comment by nathanhammond** [comment]
For those of you tagged on this issue I'm curious what problems you're trying to solve by having engine discovery:
- Organizational?
- Deploy process?
- Caching?
- Something else?

Organizational problems are hard, but tractable. Writing technology to work around human problems seems like creating technical debt (treating every line of code written as tech debt) on top of organizational debt. Being unable to arbitrarily redeploy host application assets simultaneously with an engine seems like an opportunity for improved release tooling and processes. If the organizational reason for not redeploying applications is stability I'm even less of a fan of Engine Discovery as that sounds like you're doing integration testing for the first time in production (rather than doing a build/test/deploy run of the host application).

If the concern is that deploying has a time cost then I'd encourage automation of that and there are numerous addons in the `ember-cli-deploy` ecosystem which you can use or extend to make that easy. Build times are consistently improving as we do better and smarter things with Broccoli.

Caching should be dealt with in a more fundamental way which includes string loading of JavaScript assets. That's something that we're also working on using Service Workers. Engine Discovery should not be implemented for reasons of cache stability.

The existing engine patterns–without Engine Discovery–more than cover all the use cases of one of the most complex applications built with one of the largest teams in the Ember ecosystem. That application is deployed three times a day with nothing other than a toggle switch preventing us from deploying each commit. We're also likely to be the first non-trivial application using `ember-engines` in production.

**Comment by peStoney** [comment]
We are building an embedded operating system that is using a emberjs based UI. We would like to enable the user to add functionalities to the system later, that means at build time we don't know which engines will be used so it's needed to be defined at start time. I'm not entirely sure how to categorize this but I would say it's also related to the deployment process.

**Comment by MiguelMadero** [comment]
@nathanhammond thanks for putting this together. 1 and 2 are covered at a high level on https://github.com/emberjs/rfcs/pull/141 while a pre-requisite, it's also a separate requirement. 

3. ember-cli-bundle-loader defines a [config]([ember-cli-bundle-loader](https://github.com/MiguelMadero/ember-cli-bundle-loader/blob/master/tests/dummy/config/bundles.js)) where this information can be specified, it goes beyond also to specified dependencies between packages (not only for engine assets, but also vendor). Maybe we need something similar. 

4. ... 

> Deep linking (currently solved at build time by propagating all routes to the top).

On a previous version of bundle-loader (see [v0.19 asset-service](https://github.com/MiguelMadero/ember-cli-bundle-loader/blob/v0.1.9/addon/services/lazy-loader.js#L39) and [v0.19]routing-configuration](https://github.com/MiguelMadero/ember-cli-bundle-loader/blob/v0.1.9/addon/utils/lazy-routing-configuration.js#L24)), we would patch the route information *after* loading the engine. That has a few caveats that we could address if we really wanted to go down this path. The main challenge was the dependencies between link-to and the route, which have been addressed now and `transitionTo`, with a route, not URL expected the route to be previously defined.

**Comment by mike183** [comment]
Our use case is similar to that of @peStoney's, in the sense that at build time, we wouldn't actually know which engines were available.

We're developing a CMS that has an Ember.js admin panel. Our plan is to use engines as a means of allowing developers to build plugins that extend the UI/functionality of the admin panel. As these plugins can/will be developed by third parties, it would not be possible for the host application to have prior knowledge about every possible engine at build time. It would therefore need to be capable of discovering, configuring and mounting engines at runtime.

There is a bit more discussion around our use case in #58

**Comment by mike183** [comment]
> 1. We must define an engine bundle format.
Does this suggest that it would be possible to build engines as separate packages, independent of an application?

**Comment by nathanhammond** [comment]
@peStoney @mike183 I'd rephrase both of your goals as "third-party late-binding" where the engine creator doesn't actually control the host application in any way. This is exactly the security problem I'm afraid of, except intentional. :stuck_out_tongue:

**Comment by MiguelMadero** [comment]
> How should this play with the {{mount}} keyword?

This is hard. For mount, maybe we can provide a "name" area in the "shell" where engines can add routes to. That would solve the route definition, but then we still need a way to link into it. One option is to have "dynamic" links, for example, we have a dashboard, that creates links to certain routes, those routes come from server side and can link to engines that were not known at the time the dashboard was originally created, so it's not really defined on the "engine" or the "shell" app. 
In other frameworks, like [Microsoft's Smart Client Software Factory](https://msdn.microsoft.com/en-us/library/ff699185.aspx), the framework provides extension points for "engines" (in their terminology modules), those were useful to for example to add links to a navbar. The engine has to do it since only the engine knows about its routes. In our world, to avoid introducing any more concepts like MS did, this could simply be a service that contains a list of links defined by the shell app. Unfortunately, this poses another challenge with lazy loading since this information will now have to be available at the time the links are rendered.

**Comment by MiguelMadero** [comment]
>Moving toward a solution to this problem should focus initially on getting us to lazy-loading, but we should keep this in the back of our minds so as not to paint ourselves into a corner.

Thanks for putting this together. I agree, we can focus initially on lazy-loading and still keep this scenario in mind.

**Comment by knownasilya** [comment]
@nathanhammond for my use case, it's more of an internal addon system, which we'd use in our product to provide additional features beyond the base product. The engines would all be written by our team, but would be enabled per-user of our product. Some might be paid addons, while others just optional that the user could enable themselves.

**Comment by peStoney** [comment]
@nathanhammond I understand the basic security concerns. These could be addressed by loading from trusted sources only. In our case we will only allow correctly signed engines to be installed on the system and we do the signing so we're also responsible for the integrity of the engines. I think going with a trusted zone approach might be sufficent for most of the use cases (?)

**Comment by mike183** [comment]
@nathanhammond Am I correct in thinking that the security concerns that you're mentioning are that it may become possible for engines to be unintentionally mounted into an application?

If so, I don't think that it would be unreasonable to only allow engines to be discovered and mounted from trusted domains as @peStoney suggests.

**Comment by chadhietala** [comment]
Being able to independently deploy engines creates a whole slew of problems that I think are out of the scope of ember engines, but if you want to try to solve that problem there shouldn't be anything from blocking you.

To enumerate the problems that I am talking about as I did a write up internally on why this is very challenging problem:

### Issues With Autonomy When Building An App

__Note: None of these issues are Ember specific__

Majority of these issues arise in dealing with dependencies and compatibilities. You do not have such issues in backend distributed systems largely due having standardized networking protocols. I will enumerate the issues:

#### Shared Dependencies
If you are going to deploy portions of an application you must have a system in place that reconciles the fact that applications have shared dependencies. Let's say you have an application called Saturn, and you want to deploy the Messaging bundle at a different cadence than the Search bundle. Since the teams have the autonomy, Messaging adds a for-bar library at 1.0.1 to their bundle. Now Search decides it needs foo-bar at 1.0.1 as well. Here’s where we have problem, do you bundle foo-bar library with each bundle? Do you nominate it to a Shared bundle that is guaranteed to be there at run time? Do you create a “race” and whatever bundle loads first brings along the dependency? To solve this issue the build system must be aware of the dependency graph to make the most optimal bundling decision.

#### Versioning
To make matters more complicated take the above scenario and change the fact that Messaging wants version 2.0.3 and Search wants 1.0.1 of foo-bar. Now do you not only have network issues as you might need to send 2 versions of the same library, you now need to ensure that at runtime the right bundle is talking to the right version of the library. In general this means at build time you would have to create a sandboxed version of the library and re-write the AST of all the code that pointed to it.

```
import fooBar from 'foo-bar';
```

Would need to transform into:

```
import fooBar from 'foo-bar@1.0.1';
```

We have done some work in this area with [petal](https://github.com/stefanpenner/petal). Furthermore, you would need to integrate some sort of dead code elimination process like rollup to ensure multiple versions of the same library are as small as possible.

#### Runtime Dependencies
To make matters even more complicated dependency injection systems make this much harder to solve with static analysis and would require human intervention. Ember specifically requires [runtime contracts](https://github.com/dgeb/ember-engines#declaring-dependencies) as a means to know about injections.

**Comment by chadhietala** [comment]
These types of problems should be solved more generically in Ember CLI and not a divergent solution for Engines.

**Comment by sglanzer** [comment]
The dependency issue absolutely appears to be one of the nastiest problems to solve

I just wanted to throw our use case on the pile as well - we're building a network managemen system on a microservice architecture and the UI will effectively be required to add/remove features depending on what a particular customer puchases or activates at a given time.  No addition, we want the ability for customers or 3rd party developers (e.g. other equipment vendors) to be able to contribute engines as a means of customization/extension of capabilites.

It's likely that we won't be able to achieve the full vision of this for a while, but we're pushing hard on current engine capabilites to figure out what can be done as an initial stage.

If anyone is curious, we have our navigation system engine-ready where the engines register their navigation endpoints via a service.  https://github.com/ciena-frost/ember-frost-navigation

A few other attachment points (such as a dashboard) will also be part of the first effort.

**Comment by sglanzer** [comment]
At this point I'm expecting that we'll need to provide effectively a set of LTS addons to reduce the dependency gap (not a full solution, but it should help)

**Comment by mike183** [comment]
I've been thinking about this again recently and after spending some time before Christmas hacking around with potential solutions, I'd like to outline some of my current thoughts here so that we can try and get the ball rolling.

I managed to put together a couple of commits that introduced experimental support for the dynamic discovery of engines:

- status200/ember-engines@d5fdc8c1dba6357cfca62cdcaa5077dd3f0051f0 - Separates the engines `routes.js` file (including its dependencies) from the hosts vendor file, instead we want it to be packaged into a standalone file to be included separately
- status200/ember-engines@4a986ea6c275c6c4d3620e9297fc512607c27c41 - Checks the asset manifest for any bundles containing a `routes.js` file and ensures that they are included at runtime, application initialization is deferred until the `routes.js` files are loaded

Something you may notice from the first commit is that a new option (`dynamicallyLoaded`) has been introduced. While the name is up for debate, this option tells the lazy loading build process that we don't want the engines `routes.js` file to be packaged into the hosts vendor file.

Although, instead of adding the new option alongside the current `lazyLoaded` option, I think a better approach would be allow `lazyLoaded` to accept an object as well as a boolean value. Not only would this give the new `dynamicallyLoaded` option more context, it would also provide a place for future options to be added that would allow for further customization of the lazy loading build process.

E.g

```js
lazyLoaded: {
  enabled: true,
  dynamicallyLoaded: true
}
```

Anyones thoughts on this approach would be appreciated. If this is an acceptable approach, I'm happy to put in the work to cleanup/improve my current work, add tests and submit a PR.

/cc @trentmwillis @marcemira
